### ISSUE TITLE
Error when loading an Autoencoder Neural Network (More than one input is Const op in function)

### ISSUE THREAD

**Issue by AndrewMartchenko** [opencv/opencv#22423]
##### System information (version)
- Operating System => Windows 10 64 Bit
- Python => 3.8.12
- OpenCV => 4.6.0
- Tensorflow => 2.9.1

And

- Operating System => Ubuntu 22.04
- Compiler=> GCC 11.2.0 x86_64
- OpenCV => 4.5.5

### Detailed description
I get the following error message when I try to read a frozen model that was trained using Tensorflow.

```
cv2.dnn.readNetFromTensorflow(frozen_model_path)
cv2.error: OpenCV(4.6.0) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\tensorflow\tf_importer.cpp:2824: error: (-2:Unspecified error) More than one input is Const op in function 'cv::dnn::dnn4_v20220524::`anonymous-namespace'::TFImporter::getConstBlob'
```

The model was trained on my Windows machine using python. In Python I am using the following code to read the frozen graph:
```python
model = cv2.dnn.readNetFromTensorflow("./frozen_graph.pb")
```

I also tried loading the model in C++ on my Ubuntu machine using the following code:
```c++
cv::dnn::dnn4::v20220524::Net model = cv::dnn::readNetFromTensorflow("./frozen_graph.pb");
```

In both cases, I the same error message.

### Example
The following python code can be used to reproduce the problem. The error occurs on the last line. I have also attached a frozen graph that was generated by this code.

The neural network in the example takes an image as input and returns and image as output. The inputs are images containing a white line on black background plus salt noise. The output is an image mask over region containing the white line. 

The code consists of the following steps:
1. Generates 1000 noisy and noise free images for training. 
    - The input data `x_data` are noisy images of lines.
    - The target data `y_data` are the corresponding noise free images.
2.  Splits the data into train and validation batches.
3. Creates and trains the model. The number of epochs is set to 10, so the results are quite poor. To get decent results, increase number of epochs to >100. 
4. Saves the model and freezes the graph.
5. Displays the prediction for the first 5 validation images.
6. Tries to load the frozen graph using opencv and fails (last line).

```python
import os
import numpy as np
import cv2
import random
from sklearn.model_selection import train_test_split
from tensorflow import keras
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Conv2D, MaxPooling2D,  Lambda, MaxPooling2D, UpSampling2D, concatenate
from tensorflow.keras.regularizers import l2
from tensorflow.keras import backend as K
from tensorflow.python.framework.convert_to_constants import convert_variables_to_constants_v2
import tensorflow as tf

IMAGE_WIDTH = 128


def random_line(width):
    rad = 2*np.pi*random.random()
    x = random.randint(0, width - 1)
    y = random.randint(0, width - 1)
    dx = np.cos(rad)
    dy = np.sin(rad)
    x0 = round(x-2*width*dx)
    y0 = round(y-2*width*dy)
    x1 = round(x+2*width*dx)
    y1 = round(y+2*width*dy)
    return (x0, y0), (x1, y1)


def add_noise(img):
    row, col = img.shape[:2]
    indices = np.random.choice(row*col, replace=False, size=int(row*col*0.25))
    img[np.unravel_index(indices, (row, col))] = 255
    return img


def create_data(n):
    print("Creating data")
    x_data = []
    y_data = []
    for i in range(n):
        p0, p1 = random_line(IMAGE_WIDTH)
        y_image = np.zeros((IMAGE_WIDTH, IMAGE_WIDTH, 3), dtype=np.uint8)
        cv2.line(y_image, p0, p1, (255, 255, 255), 1)
        x_image = y_image.copy()
        x_image = add_noise(x_image)
        x_data.append(x_image)
        y_data.append(y_image)
    return np.uint8(x_data), np.uint8(y_data)


def dice_coef(y_true, y_pred):
    multed = K.sum(y_true*y_pred, axis=(0, 1, 2))
    summed = K.sum(y_true+y_pred, axis=(0, 1, 2))
    return K.mean((2. * multed) / summed)


def dice_coef_loss(y_true, y_pred):
    return 1 - dice_coef(y_true, y_pred)


def new_model():
    conv2D_args = {
        'activation': 'elu',
        'padding': 'same',
        'kernel_initializer': 'glorot_uniform',
        'kernel_regularizer': l2(1e-4),
    }
    pool2D_args = {
        'pool_size': (2, 2),
        'padding': 'same',
    }
    inputs = Input((IMAGE_WIDTH, IMAGE_WIDTH, 3),
                   dtype=tf.uint8, name='input')
    std1 = Lambda(lambda x: tf.cast(x, tf.float32)/255.0-0.5)(inputs)
    # REDUCE
    conv1 = Conv2D(16, 5, **conv2D_args)(std1)
    pool1 = MaxPooling2D(**pool2D_args)(conv1)
    conv2 = Conv2D(32, 3, **conv2D_args)(pool1)
    pool2 = MaxPooling2D(**pool2D_args)(conv2)
    conv3 = Conv2D(64, 3, **conv2D_args)(pool2)
    pool3 = MaxPooling2D(**pool2D_args)(conv3)
    conv4 = Conv2D(64, 3, **conv2D_args)(pool3)
    pool4 = MaxPooling2D(**pool2D_args)(conv4)
    conv5 = Conv2D(64, 3, **conv2D_args)(pool4)
    # EXPAND
    up4 = Conv2D(64, 3, **conv2D_args)(UpSampling2D(size=(2, 2))(conv5))
    concat4 = concatenate([conv4, up4], axis=3)
    conv4a = Conv2D(64, 3, **conv2D_args)(concat4)
    up3 = Conv2D(64, 3, **conv2D_args)(UpSampling2D(size=(2, 2))(conv4a))
    concat3 = concatenate([conv3, up3], axis=3)
    conv3a = Conv2D(64, 3, **conv2D_args)(concat3)
    up2 = Conv2D(32, 3, **conv2D_args)(UpSampling2D(size=(2, 2))(conv3a))
    concat2 = concatenate([conv2, up2], axis=3)
    conv2a = Conv2D(32, 3, **conv2D_args)(concat2)
    up1 = Conv2D(16, 5, **conv2D_args)(UpSampling2D(size=(2, 2))(conv2a))
    concat1 = concatenate([conv1, up1], axis=3)
    conv1a = Conv2D(16, 5, **conv2D_args)(concat1)
    conv10 = Conv2D(3, 1, activation='sigmoid', name='output',)(conv1a)

    model = Model(inputs=inputs, outputs=conv10)
    opt = keras.optimizers.Adam(lr=5e-5)
    model.compile(optimizer=opt, loss=dice_coef_loss,
                  metrics=[dice_coef_loss, ])
    model.summary()
    return model


def freeze_model(model, frozen_graph_path):

    frozen_graph_dir, frozen_graph_filename_ext = os.path.split(
        frozen_graph_path)
    frozen_graph_filename = os.path.splitext(frozen_graph_filename_ext)[0]
    full_model = tf.function(lambda x: model(x))
    full_model = full_model.get_concrete_function(
        tf.TensorSpec(model.inputs[0].shape, model.inputs[0].dtype))
    frozen_func = convert_variables_to_constants_v2(full_model)
    frozen_func.graph.as_graph_def()
    tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
                      logdir=frozen_graph_dir,
                      name=f"{frozen_graph_filename}.pb",
                      as_text=False)
    tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
                      logdir=frozen_graph_dir,
                      name=f"{frozen_graph_filename}.pbtxt",
                      as_text=True)


if __name__ == '__main__':
    print(cv2.__version__)
    print(tf.__version__)
    EPOCHS = 10
    TRAIN = True

    tf_model_path = './tf_model/'
    frozen_model_path = './frozen_model/frozen_graph.pb'

    x_data, y_data = create_data(1000)
    x_train, x_val, y_train, y_val = train_test_split(
        x_data, y_data, train_size=0.7, random_state=42)

    if TRAIN:
        y_train_norm = y_train.astype(np.float32)/255.0
        y_val_norm = y_val.astype(np.float32)/255.0

        model = new_model()

        model.fit(x_train, y_train_norm, validation_data=(x_val, y_val_norm),
                  batch_size=32, epochs=10, shuffle=True, verbose=2)

        model.save(tf_model_path)
        freeze_model(model, frozen_model_path)

    model = keras.models.load_model(tf_model_path, custom_objects={
        'tf': tf, 'dice_coef_loss': dice_coef_loss})

    y_val_pred = model.predict(x_val)
    for i in range(5):
        x_image = x_val[i]
        y_image = (y_val_pred[i]*255.0).astype(np.uint8)
        cv2.namedWindow('x', cv2.WINDOW_NORMAL)
        cv2.namedWindow('y', cv2.WINDOW_NORMAL)
        cv2.resizeWindow('x', 400, 400)
        cv2.resizeWindow('y', 400, 400)
        cv2.moveWindow('x', 0, 0)
        cv2.moveWindow('y', 410, 0)
        cv2.imshow('x', x_image)
        cv2.imshow('y', y_image)
        k = cv2.waitKey(0)
        if k == ord('q'):
            break
    cv2.destroyAllWindows()

    # Loading frozen graph causes error
    cv2.dnn.readNetFromTensorflow(frozen_model_path)

```
[frozen_graph.zip](https://github.com/opencv/opencv/files/9413819/frozen_graph.zip)
`

**Comment by AndrewMartchenko** [comment]
The frozen model was not attached correctly in the original issue and github is not letting me fix it. Here is the frozen model again
[frozen_graph.zip](https://github.com/opencv/opencv/files/9413899/frozen_graph.zip)
.

**Comment by OCI1** [comment]
I am have also recently experienced the same issue with other code and upon running the code here the error is replicated. I have tried the code with the same versions as stated:
    Operating System => Windows 10 64 Bit
    Python => 3.8.12
    OpenCV => 4.6.0
    Tensorflow => 2.9.1
I have also tried it with: 
    Operating System => Windows 10 64 Bit
    Python => 3.10
    OpenCV => 4.5.5
    Tensorflow => 2.8.0

I experience the same issue's on both!

**Comment by AndrewMartchenko** [comment]
I have managed to isolate the issue to the UpSampling2D layers. For example, if we connect conv1 layer directly to conv10, then the I can load the frozen graph with `cv2.dnn.readNetFromTensorflow(frozen_model_path)`.

```python
def new_model():
    conv2D_args = {
        'activation': 'elu',
        'padding': 'same',
        'kernel_initializer': 'glorot_uniform',
        'kernel_regularizer': l2(1e-4),
    }
    pool2D_args = {
        'pool_size': (2, 2),
        'padding': 'same',
    }
    inputs = Input((IMAGE_WIDTH, IMAGE_WIDTH, 3),
                   dtype=tf.uint8, name='input')
    std1 = Lambda(lambda x: tf.cast(x, tf.float32)/255.0-0.5)(inputs)
    # REDUCE
    conv1 = Conv2D(16, 5, **conv2D_args)(std1)
    conv10 = Conv2D(3, 1, activation='sigmoid', name='output',)(conv1)

    model = Model(inputs=inputs, outputs=conv10)
    opt = keras.optimizers.Adam(lr=5e-5)
    model.compile(optimizer=opt, loss=dice_coef_loss,
                  metrics=[dice_coef_loss, ])
    model.summary()
    return model
```

I have also ruled out the concatenate layers, because when I remove them from the network I still get the same error message.

Strange thin is, this code worked about 6 months ago, but I cannot figure out what version of OpenCV of Tensorflow I was using back then.

Has anyone having problems with UpSampling2D?

**Comment by AndrewMartchenko** [comment]
I managed to get the model loaded in OpenCV by converting the tensorflow model to an ONNX format.
